File: backend/models/alert.py
import uuid
from datetime import datetime
from sqlalchemy import String, Boolean, Text, DateTime, func, ForeignKey, JSON
from sqlalchemy.orm import Mapped, mapped_column
from models.database import Base


class Alert(Base):
    __tablename__ = "alerts"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    event_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    time: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), index=True)
    resident_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("residents.id", ondelete="SET NULL"), nullable=True)
    severity: Mapped[str] = mapped_column(String(20), nullable=False)
    alert_type: Mapped[str] = mapped_column(String(100), nullable=False)
    category: Mapped[str] = mapped_column(String(50), nullable=False)
    zone: Mapped[str] = mapped_column(String(100), nullable=False)
    message: Mapped[str] = mapped_column(Text, nullable=False)
    acknowledged: Mapped[bool] = mapped_column(Boolean, default=False)
    acknowledged_by: Mapped[str | None] = mapped_column(String(200))
    acknowledged_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    resolved: Mapped[bool] = mapped_column(Boolean, default=False)
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    notification_sent: Mapped[bool] = mapped_column(Boolean, default=False)
    metadata_: Mapped[dict] = mapped_column("metadata", JSON, default=dict)

File: mobile/lib/models/alert.dart
class SentinelAlert {
  final String id;
  final DateTime time;
  final String severity;
  final String alertType;
  final String category;
  final String zone;
  final String message;
  final bool acknowledged;
  final String? residentId;

  const SentinelAlert({
    required this.id,
    required this.time,
    required this.severity,
    required this.alertType,
    required this.category,
    required this.zone,
    required this.message,
    required this.acknowledged,
    this.residentId,
  });

  factory SentinelAlert.fromJson(Map<String, dynamic> j) => SentinelAlert(
        id: j['id'],
        time: DateTime.parse(j['time']),
        severity: j['severity'],
        alertType: j['alert_type'],
        category: j['category'],
        zone: j['zone'],
        message: j['message'],
        acknowledged: j['acknowledged'] ?? false,
        residentId: j['resident_id'],
      );

  String get severityEmoji {
    switch (severity) {
      case 'critical': return '🚨';
      case 'high': return '⚠️';
      case 'medium': return '🔔';
      default: return 'ℹ️';
    }
  }
}
